cocos/platform/android/jni/JniHelper-inl.h

```diff
@@ -0,0 +1,494 @@
+﻿/****************************************************************************
+ Copyright (c) 2010-2012 cocos2d-x.org
+ Copyright (c) 2015 Victor Komarov
+ Copyright (c) 2017 Jeff Wang
+
+ http://www.cocos2d-x.org
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+#ifndef __ANDROID_JNI_HELPER_INL_H__
+#define __ANDROID_JNI_HELPER_INL_H__
+
+#include <type_traits>
+#include "base/ccUTF8.h"
+
+NS_CC_BEGIN
+
+template <class Ty> struct CStyleArrayWrapper {
+    explicit CStyleArrayWrapper(const Ty *e, size_t s) : elem(e), size(s) { }
+    const Ty *elem;
+    size_t size;
+};
+
+template <class Ty>
+inline CStyleArrayWrapper<Ty> makeCStyleArray(const Ty *e, size_t s) {
+    return CStyleArrayWrapper<Ty>(e, s);
+}
+
+//
+// LocalRefWrapper
+//
+struct LocalRefWrapper {
+    explicit LocalRefWrapper(JNIEnv* env, jobject obj) : _env(env), _obj(obj) { }
+    ~LocalRefWrapper() { _env->DeleteLocalRef(_obj); }
+    inline jobject get() const { return _obj; }
+
+private:
+    LocalRefWrapper(const LocalRefWrapper&) = delete;
+    LocalRefWrapper(LocalRefWrapper&&) = delete;
+    LocalRefWrapper& operator=(const LocalRefWrapper&) = delete;
+    LocalRefWrapper& operator=(LocalRefWrapper&&) = delete;
+
+protected:
+    JNIEnv* _env;
+    jobject _obj;
+};
+
+namespace JniHelperDetail {
+
+    //
+    // CharSequence
+    //
+    template <char... Chars>
+    struct CharSequence {
+        static const char value[sizeof...(Chars) + 1];
```

Comment:
Why need static variable?

Reply:
CharSequence::value is used for deduct JNI signature during compiling time

Reply:
What i mean is that, can change to non static variable, use member variable instead, or these memory can not be freed.

Reply:
as a rule of thumb, template metaprogram uses static member.I will have a try to modify it as a non static variable